compiler/rustc_codegen_ssa/src/debuginfo/type_names.rs

```diff
@@ -202,17 +294,14 @@ pub fn push_debuginfo_type_name<'tcx>(
             // processing
             visited.remove(t);
         }
-        ty::Closure(def_id, ..) => {
-            output.push_str(&format!(
-                "closure-{}",
-                tcx.def_key(def_id).disambiguated_data.disambiguator
-            ));
-        }
-        ty::Generator(def_id, ..) => {
-            output.push_str(&format!(
-                "generator-{}",
-                tcx.def_key(def_id).disambiguated_data.disambiguator
-            ));
+        ty::Closure(def_id, ..) | ty::Generator(def_id, ..) => {
+            let key = tcx.def_key(def_id);
+            if qualified {
+                let parent_def_id = DefId { index: key.parent.unwrap(), ..def_id };
+                push_item_name(tcx, parent_def_id, true, output);
```

Comment:
While chatting with @wesleywiser about this it occurred to me that the generic parameters of a closure don't show up anywhere in the type name. So, if you have a generic function with a closure like
```rust
fn foo<T>(x: T) -> (T, u32) {
    let some_closure = |y: T| (y, 0);
    some_closure(x)
}
```
and then multiple instantiations of the function then the two closure instances 
```rust
pub fn bar() -> u32 {
    foo(0u16).1 + foo(0u32).1
}
``` 
then the all closure instances would show up as `foo::closure$0`. I think the best way to handle this would be to add the type parameters to the enclosing function, as in `foo<u16>::closure$0`, `foo<u32>::closure$0`, etc.

(see https://godbolt.org/z/WGrhvs81e for the example)

Reply:
I added a test for this, and it look like the generic parameters are maintained but they appear on the closure instead of the declaring function, so for your example it would appear as `foo::closure$0<u16>`